### DISCUSSION TITLE
implementation of reflection-data output

### DISCUSSION THREAD

**Discussion by a-stana** [sam-xl/scancontrol#40] [General]
Hi all,
it would be very appreciated if the implementation of reflection data in the output could be considered in the more or less near future. It would be very interesting for my current work. I will try my best to contribute to this GitHub specifically in this direction.

**Comment by dave992** [comment]
This is indeed currently not implemented in the driver, it only relays the points captured from the first reflection. Currently, I am using a PointCloud XYZI format, which is not suitable for containing the other reflections. Do you have any ideas on how this can be implemented instead? 

Out of curiosity, for what kind of application would the other reflections have a benefit? From my experience with the sensor, there is a very limited amount of data available in the 2nd to 4th reflection.

**Comment by a-stana** [comment]
We want to develop an autonomous robot for precision crop sowing and scan the soil surface to estimate seeding depth. For this task XYZPointclouds is sufficient.
Another task could be to scan the furrow with the crop inside, here the reflection data would be more reliant to distinguish seeds from soil. I ran some pre-trials and found that the reflection-width looked most promising. 
I'm quite new to point-cloud-data but is it may be possible to implement the data into a modified XYZRBG? Although I know that the values of 0-255 are not enough in size for keeping reflection-width m0 and intensity data. Or one could maybe save 2 point-clouds, one containing the XYZ coordinates and another the reflection values instead. I'm not sure whether this is a possibility.

**Comment by gavanderhoorn** [comment]
For pure laser scanners there is the [MultiEchoLaserScan](http://docs.ros.org/en/api/sensor_msgs/html/msg/MultiEchoLaserScan.html) message which could be used, but I believe that is not a good idea here as it models a sensor with an origin an polar coordinate system. The Scan Control devices do not fall in that category.

> Currently, I am using a PointCloud XYZI format, which is not suitable for containing the other reflections

A work-around could be to "store multiple clouds in a single one". This would make it harder to figure out which points come from the additional reflections, but would support as many reflections as we'd like, as they'd just translate into more points in the same cloud.

A custom message (carrying multiple `sensor_msgs/PointCloud2` messages) could also be used, but should perhaps be the last resort, as it will guarantee incompatibility with existing tools.

It might be good to search for other devices which support multiple returns and have ROS drivers. Then analyse how the authors/maintainers of those components have approached the problem.

Would that be something you could do @a-stana?

---

Edit: it is possible to store 'custom' fields in a `PointCloud2` message. Compatibility with existing tooling would perhaps be limited though.

**Comment by a-stana** [comment]
ok, I'll try to do that, but it could take a few days! I also checked the SDK 0.2.4 documentation again which says that intensity and reflection-width are both 10-bit values, where m0 and m1 are represented by 32-bit values. Wouldn't it be possible to put out another XYZI-PCL with the reflection-width instead of intensity since the values are the same in size?

**Comment by dave992** [comment]
>  Wouldn't it be possible to put out another XYZI-PCL with the reflection-width instead of intensity since the values are the same in size?

The length and position of the data are presented in the SDK documentation so you are able to extract the information you are interested in from the buffer (for which a function `ConvertPartProfile2Values()` is provided actually) and for determining the  `partial_profile` settings if you want to reduce the amount of data sent. It is not a limitation for the `point_cloud` messages (where the intensity is represented as a `float`).  

Just to clarify, are you interested in the `reflection width` data or in the different reflections/peaks (or both)? If you want to only add the `reflection width` data, it might be best to add a custom PointField (as mentioned by gavanderhoorn). It just has to be investigated if this breaks compatibility with existing tools. 

*A small note, the intensity is actually not used at the moment (It is always 1.0 in the current driver version) as the default settings use a `partial_profile` where only the x/z data is sent for the first peak/reflection. It is quite trivial to add in the actual intensity if needed.

**Comment by a-stana** [comment]
Well,  I'm interested in all the information long-term but, to be honest, right now the width would be sufficient. Nevertheless, @gavanderhoorn I will keep an eye out for other laser solutions and their ROS drivers. And try to play around with switching on the recording of reflection-intensity and/or replacing it with reflection-width. Thank you guys so far. I'll report on my outcome.
